src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,260 @@
+        """
+        TODO(meredith): add doc string
+        """
+        mapped_results: MutableMapping[int, MutableMapping[str, int]] = {}
+
+        cache_keys = KeyCollection(org_strings)
+
+        # Step 1: Lookup ids in the cache and add them to the
+        # cache KeyCollection results.
+        cache_hits = 0
+        cache_misses = 0
+        cache_key_strs = cache_keys.as_strings()
+        cache_results = indexer_cache.get_many(cache_key_strs)
+        for cache_key in cache_key_strs:
+            result = cache_results.get(cache_key)
+            org_id, string = cache_key.split(":")
+            if result:
+                cache_keys.add_key_result((int(org_id), string), result)
+                cache_hits += 1
+            else:
+                cache_misses += 1
+
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "true", "caller": "get_many_ids"},
+            amount=cache_hits,
+        )
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "false", "caller": "get_many_ids"},
+            amount=cache_misses,
+        )
+        mapped_results.update(cache_keys.get_mapped_results())
+
+        db_read_keys = cache_keys.get_unmapped_keys()
+
+        if not db_read_keys.as_strings():
+            return mapped_results
+
+        # Step 2: Lookup ids in the db and add them to the
+        # db_read KeyCollection results.
+        for db_obj in self._get_db_records(db_read_keys):
+            db_read_keys.add_key_result((db_obj.organization_id, db_obj.string), db_obj.id)
+
+        db_read_key_results = db_read_keys.get_mapped_results()
+
+        if len(db_read_key_results.keys()) > 0:
+            results_to_cache = db_read_keys.get_mapped_key_strings_to_ints()
+            indexer_cache.set_many(results_to_cache)
+            mapped_results.update(db_read_key_results)
+
+        db_write_keys = db_read_keys.get_unmapped_keys()
+
+        if not db_write_keys.as_strings():
+            return mapped_results
+
+        # Step 3: Create new records in the db and add them to the
+        # db_write KeyCollection results.
+        new_records = []
+        for write_pair in db_write_keys.as_tuples():
+            organization_id, string = write_pair
+            new_records.append(
+                StringIndexerTable(organization_id=int(organization_id), string=string)
+            )
+
+        with metrics.timer("sentry_metrics.indexer.pg_bulk_create"):
+            # We use `ignore_conflicts=True` here to avoid race conditions where metric indexer
+            # records might have be created between when we queried in `bulk_record` and the
+            # attempt to create the rows down below.
+            StringIndexerTable.objects.bulk_create(new_records, ignore_conflicts=True)
+
+        for db_obj in self._get_db_records(db_write_keys):
+            db_write_keys.add_key_result((db_obj.organization_id, db_obj.string), db_obj.id)
```

Comment:
However you will design the data structure that holds the mapping results, it seems that you are always updating that in batches, so you will make the code much more readable if you provided an `add_key_results` method that takes a sequence of mapping (or a dict, depending how you want to design it) instead of iterating on the results you got from the DB and adding them one by one.